### DISCUSSION TITLE
Provide an API endpoint to submit offline listens

### DISCUSSION THREAD

**Discussion by Maxr1998** [jellyfin/jellyfin-meta#61] [Ideas]
At the moment, the playback session API doesn't allow you to submit sessions from the past, so playback reporting plugins like the ListenBrainz and Last.fm plugin have no (proper) way to record such listens.<sup>1</sup>

A long time suggestion has been to add a timestamp to the [ReportPlaybackStopped API](https://api.jellyfin.org/#tag/Playstate/operation/ReportPlaybackStopped)
But because this would require a separate request for every recorded offline listen, I don't think this would be a good solution.
Instead, clients should be able to submit a list of listens once they are online again, which could then be submitted to the activity log and be broadcast to playback reporting plugins so that they could pick them up and submit them to an external service.

By making this an official API, behavior across various plugins could be streamlined and clients would only need to use the Jellyfin API and not need to bother with external services or third-party APIs.
Moreover, an official API should hopefully increase adoption among third-party music playback clients.

I've discussed this with the ListenBrainz plugin developer already and posted some further thoughts in lyarenei/jellyfin-plugin-listenbrainz#51.

There's also a draft for the API design in that ticket. I'd love to discuss the API proposition here further, and afterward work on getting this into the server, if accepted.

[1] The ListenBrainz plugin offers an [alternative mode](https://github.com/lyarenei/jellyfin-plugin-listenbrainz/blob/main/doc/configuration.md#userdatasaved-event-mode) that does actually allow to retroactively submit listens, however, it comes with its own share of disadvantages as outlined in the docs above, and also requires one request per reported offline playback.

**Comment by thornbill** [comment]
This seems useful for any app that supports (or plans to support) offline playback... including the [Expo app](https://github.com/jellyfin/jellyfin-expo/issues/372).

**Comment by Chaphasilor** [comment]
I'm immensely in favor of such an endpoint. Most Jellyfin music clients that support playback reporting focus on offline playback (Finamp, Symfonium), and from what I've seen there are quite a few users that listen in offline mode a lot. So a way to submit offline listens does make a lot of sense, otherwise what's the point of playback reporting if the data is incomplete?

Given that I'm also creating Jellyfin Rewind, I'm always interested in extensive and complete playback stats :)
